reconcile/openshift_base.py

```diff
@@ -345,11 +345,47 @@ def apply(
         except FieldIsImmutableError:
             # Add more resources types to the list when you're
             # sure they're safe.
-            if resource_type not in ["Route", "Service", "Secret"]:
+            if resource_type not in ["Route", "Service", "Secret", "Deployment"]:
                 raise
 
-            oc.delete(namespace=namespace, kind=resource_type, name=resource.name)
-            oc.apply(namespace=namespace, resource=annotated)
+            if resource_type == "Deployment":
+                logging.info(
+                    ["replace", cluster, namespace, resource_type, resource.name]
+                )
+                # spec.selector changes
+                current_resource = oc.get(namespace, resource_type, resource.name)
+
+                # check update strategy
+                if current_resource["spec"]["strategy"]["type"] != "RollingUpdate":
+                    logging.error(
+                        f"Can't replace Deployment '{resource.name}' inplace w/o interruption because spec.strategy.type != 'RollingUpdate'"
+                    )
+                    raise
+
+                # Get active ReplicSet for old Deployment. We've to delete it manually after
+                # the new Deployment is in place.
+                obsolete_rs = oc.get_replicaset(namespace, current_resource)
+
+                # delete old Deployment
+                oc.delete(
+                    namespace=namespace,
+                    kind=resource_type,
+                    name=resource.name,
+                    cascade=False,
+                )
+                # create new one
+                oc.apply(namespace=namespace, resource=annotated)
+                if obsolete_rs:
+                    oc.wait_for_deployment(namespace, resource.name)
+                    # delete old ReplicSet
+                    oc.delete(
+                        namespace=namespace,
+                        kind=obsolete_rs["kind"],
+                        name=obsolete_rs["metadata"]["name"],
+                    )
```

Comment:
i would have expected that the new deployment takes ownership of the old ReplicaSet and handles scaledown of its pods during the RollingUpgrade. if that is true (i honestly don't know for sure), the deletion of the replicaset would not be necessary. i'm happy to learn something new here

Reply:
after discussing with @chassing i learned that the ownership reference of the old ReplicaSet does not only include name and kind but also the uid, which changes when the Deployment is recreated.

idea: after replacing the Deployment, update the old ReplicaSet ownership reference with the uid of the new Deployment. this way no waiting for Deployment success and no deletion of the old ReplicaSet "should" be necessary. worth testing because it would imply to get rid of the waiting.

Reply:
I tried it, but it doesn't work :( After deletion of the old Deployment, the `ownerReference` disappeared in the old ReplicaSet; trying to add it again, but it was removed immediately.

Reply:
there is a way to do it. when you update the old `ReplicaSet` with the `ownerReference` (with `controller: true`), you also need to update the `metadata.labels` of it to match the `spec.selector.matchLabels` of the `Deployment` (mins the `pod-template-hash`). this is the roughly the way operator SDK does it behind the scenes in its adoption logic. 🤯 

it is a lot more effort but could be still worth it in order to get rid of the waiting.

it is completely up to you if you want to go that road.

Reply:
Yeah, that's working like a charm. I'll adapt the code.

Reply:
@geoberle PTAL

Reply:
nice. LVGTM